--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -905,8 +905,28 @@
 				    tx_pause, rx_pause);
 }
 
+static void mtk_validate(struct phylink_config *config,
+                         unsigned long *supported,
+                         struct phylink_link_state *state)
+{
+        __ETHTOOL_DECLARE_LINK_MODE_MASK(mask) = { 0, };
+
+        phylink_set(mask, Pause);
+        phylink_set(mask, Asym_Pause);
+
+        phylink_set(mask, 10baseT_Half);
+        phylink_set(mask, 10baseT_Full);
+        phylink_set(mask, 100baseT_Half);
+        phylink_set(mask, 100baseT_Full);
+        phylink_set(mask, 1000baseT_Full);
+        phylink_set(mask, 2500baseT_Full);
+
+        linkmode_and(supported, supported, mask);
+        linkmode_and(state->advertising, state->advertising, mask);
+}
+
 static const struct phylink_mac_ops mtk_phylink_ops = {
-	.validate = phylink_generic_validate,
+	.validate = mtk_validate,
 	.mac_select_pcs = mtk_mac_select_pcs,
 	.mac_config = mtk_mac_config,
 	.mac_finish = mtk_mac_finish,
